Blog: Protect mission-critical pods with PriorityClasses (#37713)
kubernetes/website: content/en/blog/_posts/protect-mission-critical-pods-priorityclass/index.md

---
layout: blog
title: "Protect your mission-critical pods from eviction with PriorityClass"
date: 2022-11-04
slug: protect-mission-critical-pods-priorityclass
description: "Pod priority and preemption help to make sure that mission-critical pods are up in the event of a resource crunch by deciding order of scheduling and eviction."
---

**Author:** Sunny Bhambhani (InfraCloud Technologies)

> The adoption of Kubernetes is at a high pace now and most organizations are using this as a de-facto orchestration engine for running their workloads. And at this high pace, the workloads are created and deleted frequently.

Reviewer: ```suggestion
The adoption of Kubernetes is at a high pace now and many organizations are using this as a de-facto orchestration engine for running their workloads. At this high pace, workloads (and instances of those workloads) are created and deleted frequently.
```

> Therefore, scheduling of pods is crucial in every Kubernetes cluster, and proper scheduling of the pods is the key to running a cluster without issues and making sure that our application pods are up and running. In this article, we will talk about some of the use cases around resource management and how PriorityClass can help in protecting our mission-critical or high-priority pods from getting evicted and making sure that our application pods are up and running and serving traffic.

Reviewer: ```suggestion
Therefore, scheduling of pods is crucial in every Kubernetes cluster, and proper scheduling of the pods is the key to running a cluster without issues and making sure that our application pods are up and running. In this article, I will explain some of the use cases around resource management. I'll cover how PriorityClass can help in protecting mission-critical or high-priority pods from getting evicted and making sure that our application pods are up and running and serving traffic.
```

## Resource Management in Kubernetes 

> The controlplane consists of multiple components, out of which the scheduler (aka kube-scheduler) is one of the components which is responsible for assigning a node to a pod.

Reviewer: ```suggestion
The control plane consists of multiple components, out of which the scheduler (usually the built-in [kube-scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/)) is one of the components which is responsible for assigning a node to a pod.
```

- Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.

>   - In the background, the scheduler runs an infinite loop that looks out for all the pods where the nodeName field is null. When it finds one, it starts the process of finding which node should be the best one. Though there are various parameters for it, just to name a few, they are “nodeName”, “nodeSelector”, “taint and tolerations”, "nodeAffinity", and the rank of available nodes based on available resources in terms of CPU and memory, etc which decides the pod’s placement.

Reviewer: ```suggestion
  - In the background, the scheduler runs as an infinite loop that looks out for all the pods where the `nodeName` field is null. When it finds one, it starts the process of finding which node should be the best one. Though there are various parameters for it, just to name a few, they are `nodeName`, `nodeSelector`, taints, tolerations, `nodeAffinity`, and the rank of available nodes based on available resources in terms of CPU and memory, etc which decides the pod’s placement.
```

- If in-case the scheduler is not able to find any node, the pod remains to be in pending state, which is not an ideal scenario.

The below diagram explains how the request flows in (starting from point number 1 to 4):

> ![Whenever there is a new incoming pod, this is how kube-scheduler schedules a pod.](kube-scheduler.svg)

Author: +{{< figure src=kube-scheduler.svg alt="Whenever there is a new incoming pod, this is how kube-scheduler schedules a pod." title="Kube Scheduler">}}

## Typical Use Cases

1. Let's say the pod which you are planning to deploy is really critical, and you need to run it anyhow. How will you make the resources available in the cluster? By first identifying which pods are not required, then manually deleting them, etc., which is a manual and tedious process, or you may plan to add a new node to the cluster, which too might take some time and money as well.

> An example of such a case is a DaemonSet of an [infrastructure component like Grafana Loki](https://www.infracloud.io/blogs/grafana-loki-log-monitoring-alerting/), the Loki pods must run on each node before any other pods run on that node.

Reviewer: ```suggestion
An example of such a case is a DaemonSet of an infrastructure component like Grafana Loki. In that case, the Loki pods must run on each node before any other pods run on that node.
```

2. Another use case can be: Suppose, you have a single cluster, and that cluster holds the pods for the below environments:
   - preprod [Priority 1] 
   - staging [Priority 2] 
   - development [Priority 3]

>    Unfortunately, the cluster goes through high resource consumption, and now there is a resource crunch on the nodes with respect to the CPU and the memory, and we don’t have further nodes to expand or scale our Kubernetes cluster. Either the pods will remain in a pending state or the cluster will start behaving abnormally and if at all eviction starts, it will do that randomly because it doesn’t know which one to evict and which one to keep. 

Reviewer: ```suggestion
   Unfortunately, the cluster goes through high resource consumption, and now there is a resource competition on the nodes with respect to the CPU and the memory. At least in the short term, you don’t have further nodes to expand or scale that Kubernetes cluster (cluster level autoscaling _may_ add more nodes, but this takes time). Perhaps some pods will remain in a Pending state, or perhaps service will degrade as Pods compete for resources. If the kubelet does evict a Pod from a node, that eviction happens at random because the kubelet doesn’t have any special information about which Pods to evict and which to keep.
```

There can also be other scenarios where you want to control the order of scheduling or order of eviction of pods.

## PriorityClasses in Kubernetes

PriorityClass is a cluster-wide object/resource in a Kubernetes cluster and is a part of the scheduling.k8s.io/v1 API group. It contains a mapping of the PriorityClass name (which can be defined in priorityclass.metadata.name) and an integer value (which can be defined in priorityclass.value). This is the value that the pods get and is used for defining the priority.

Do you think Kubernetes itself uses PriorityClasses? The answer is yes. When we create a cluster by kubeadm or even in a managed Kubernetes service (for example, in AKS-Azure Kubernetes Service), PriorityClasses are widely used to safeguard the pods which are hosted on the master node, i.e., all the control-plane components/pods and other pods as well, like CoreDNS, metrics-server, kube-proxy, etc.

> PriorityClass makes sure that these pods are up and alive to serve the requests and make sure that the cluster is up and running.

Reviewer: ```suggestion
Using a special PriorityClass helps to make sure that these pods are up and alive to serve the requests and make sure that the overall cluster is up and running.
```

```console

> $ kubectl get pc

Author: +$ kubectl get priorityclass
-_Note: pc is the short name of PriorityClass._

NAME                      VALUE        GLOBAL-DEFAULT   AGE
system-cluster-critical   2000000000   false            82m
system-node-critical      2000001000   false            82m
```

_Note: pc is the short name of PriorityClass._

The below diagram shows exactly how it works and also contains the details of the example, which will come in the upcoming section (PriorityClass HandsOn Example).

> ![Whenever there is a new incoming pod, this is how kube-scheduler decides their priority and puts them in a queue](decision-tree.svg)

Author: Sure, updated with below code:

+{{< figure src=decision-tree.svg alt="Whenever there is a new incoming pod, this is how kube-scheduler decides their priority and puts them in a queue" title="Decision Tree">}}

### Pod priority and preemption

Pod preemption is a Kubernetes feature that allows the cluster to evict/preempt lower-priority pods over high-priority pods in the event of a resource crunch. To learn more about this concept, please refer to the Kubernetes [official documentation](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/).

Pod priority and preemption are Kubernetes features:
- Which help in deciding a priority while scheduling pods.
- Which check if the mission-critical pods are up and running and are not evicted.
- Decide how the cluster will behave in the event of a resource crunch.
- Decide which lower priority pods are preempted (evicted) from the nodes when there are resource crunch and high priority pods are available in the queue for scheduling.

### PriorityClass requirements

> 1. Decision on which all PriorityClasses are needed.

Reviewer: ```suggestion
Before you set up PriorityClasses, there are a few things to consider.

1. Decide on which  PriorityClasses are needed.
```

   Example: Based on environments, or based on the kind of pods, or based on the kind of applications.

> 2. A globalDefault PriorityClass resource in the cluster (otherwise the pods without priorityClassName will be treated as 0).

Reviewer: ```suggestion
2. Which PriorityClass resource will be the default for the cluster (otherwise the pods without a `priorityClassName` will be treated as priority 0).
```

3. Respective PriorityClass resources in the cluster based on the decision made on #1.

> 4. Name of the priorityClassName in the pod definition files or under respective pod templates like in ReplicaSet, Deployment, etc.

Reviewer: ```suggestion
4. How you'll name your priority classes.
5. How you'll make sure that Pods for your workloads are running with the right PriorityClass.
```

Let's see why this is happening:
```console
$ kubectl get events
...
...
5s          Warning   FailedScheduling   pod/prod-nginx      0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
```

To our surprise, our worker node is having some resource crunch.

Now, let's look at how PriorityClass can help in this situation since prod should be given higher priority than the other environments.

### PriorityClass resource

Before we create the PriorityClasses based on our requirements, let's see what a basic YAML definition looks like and some prerequisites:

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: PRIORITYCLASS_NAME
value: INT_VALUE_BETWEEN_0_TO_1000000000
description: SMALL_DESCRIPTION
globalDefault: TRUE_OR_FALSE
```

Below are some prerequisites for PriorityClasses:

- The name of a PriorityClass must be a valid DNS subdomain name.
- The name should not start with system-, the reason being that these names are already reserved by Kubernetes for its default PriorityClasses and to avoid unnecessary confusion.

> - Its value should be lower than or equal to 1 billion (1,000,000,000).

Author: Yes, I just cross-checked they are allowed, I will correct this. Thank you.

```
$ k get pc 
NAME                      VALUE         GLOBAL-DEFAULT   AGE
negative-test-pc          -1000000000   false            4s
system-cluster-critical   2000000000    false            3d11h
system-node-critical      2000001000    false            3d11h
```

```
$ k describe pod | grep -i prio
Priority:             -1000000000
Priority Class Name:  negative-test-pc
```

> - Its value should be lower than or equal to 1 billion (1,000,000,000).

Author: Corrected the statement as well.

- Larger numbers are reserved by PriorityClasses such as system-cluster-critical, system-node-critical for kube-system Pods.
- There are two optional fields:
  - _globalDefault:_ This value is used for pods where the priorityClassName is not specified and only one PriorityClass with globalDefault set to true can exist in a cluster.  
  If there is no PriorityClass defined with globalDefault set to true, all the pods with no priorityClassName defined will be treated with 0 priority (i.e. the least priority).
  - _Description:_ It's a string with a meaningful value so that people know when to use this PriorityClass.

{{< note >}}
Adding a PriorityClass with globalDefault set to true does not mean it will apply the same to the existing pods that are already running. This will be applicable only to the pods that came into existence after the PriorityClass was created.

> {{< /note >}}

Author: Honestly speaking since this was a critical point, I thought to call this out explicitly. If this doesn't look good please let me know will get this removed.

### PriorityClass in action

Now we will create some environment-specific PriorityClasses:

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: dev-pc
value: 1000000
globalDefault: false

> description: "This priority class should only be used for all development pods."

Author: Sure, 

For description used something like this `+description: >-`

```

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: preprod-pc
value: 2000000
globalDefault: false

> description: "This priority class should only be used for all pre-production pods."

Author: Sure, 

For description used something like this `+description: >-`

```

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: prod-pc
value: 4000000
globalDefault: false

> description: "This priority class should only be used for production pods."

Author: Sure, 

For description used something like this `+description: >-`

```console
$ kubectl get pods --show-labels
NAME            READY   STATUS    	RESTARTS   AGE   LABELS
dev-nginx       1/1     Terminating	0          55s   env=dev
preprod-nginx   1/1     Running   	0          55s   env=preprod
prod-nginx      0/1     Pending   	0          55s   env=prod
```

The dev-nginx pod is getting terminated. Once that is successfully terminated and we have enough resources for the prod pod, it will be scheduled.

```console
Warning   FailedScheduling   pod/prod-nginx    0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
Normal    Preempted          pod/dev-nginx     by default/prod-nginx on node node01
Normal    Killing            pod/dev-nginx     Stopping container dev-nginx
Normal    Scheduled          pod/prod-nginx    Successfully assigned default/prod-nginx to node01
Normal    Pulling            pod/prod-nginx    Pulling image "nginx"
Normal    Pulled             pod/prod-nginx    Successfully pulled image "nginx"
Normal    Created            pod/prod-nginx    Created container prod-nginx
Normal    Started            pod/prod-nginx    Started container prod-nginx
```

## Summary

> The above example and its events clearly state what this feature of Kubernetes brings to the table, plus there can be multiple other scenarios as well where we can use this feature. Just to reiterate, this helps in making sure that mission-critical pods are made up and available to serve the traffic and, in the case of a resource crunch, how the cluster should behave. 

Reviewer: ```suggestion
The above example and its events clearly state what this feature of Kubernetes brings to the table, plus there can be multiple other scenarios as well where you can use this feature. Just to reiterate, this helps in making sure that mission-critical pods are made up and available to serve the traffic and, in the case of a resource crunch, how the cluster should behave.
```

---
layout: blog
title: "Protect Your Mission-Critical Pods From Eviction With PriorityClass"

> date: 2022-12-10

Reviewer: I recommend:
```suggestion
date: 2023-01-12
```

and then update the folder name to match the pattern that other articles already follow, including that revised date.

Author: Updated the date and have modified the directory naming convention as well.

---
layout: blog
title: "Protect your mission-critical pods from eviction with PriorityClass"

> date: 2022-11-07

Author: Hi @divya-mohan0209 ,  Thanks for the review. Yes, true. Once the article is ready, we can decide on a date for it.

> date: 2022-11-07

Author: Based on the latest comment have updated the date.

- date: 2022-12-10
+ date: 2023-01-12

---

**Author:** Sunny Bhambhani (InfraCloud Technologies)

The adoption of Kubernetes is at a high pace now and many organizations are using this as a de-facto orchestration engine for running their workloads. At this high pace, workloads (and instances of those workloads) are created and deleted frequently.

Therefore, scheduling of pods is crucial in every Kubernetes cluster, and proper scheduling of the pods is the key to running a cluster without issues and making sure that our application pods are up and running. In this article, I will explain some of the use cases around resource management. I'll cover how PriorityClass can help in protecting mission-critical or high-priority pods from getting evicted and making sure that our application pods are up and running and serving traffic.

## Resource Management in Kubernetes 

The control plane consists of multiple components, out of which the scheduler (usually the built-in [kube-scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/)) is one of the components which is responsible for assigning a node to a pod.

> - Whenever a pod is created, it goes to a pending state, and then the scheduler kicks in and determines which node should be the best-suited node for the placement of the new pod.

Author: This section was added to the article so that people can relate to the content and those who are new to scheduling can get a gist of how the scheduler works.

>   - In the background, the scheduler runs as an infinite loop that looks out for all the pods where the `nodeName` field is null. When it finds one, it starts the process of finding which node should be the best one. Though there are various parameters for it, just to name a few, they are `nodeName`, `nodeSelector`, taints, tolerations, `nodeAffinity`, and the rank of available nodes based on available resources in terms of CPU and memory, etc which decides the pod’s placement.

Author: Sure, will review it and will make it simpler.

>   - In the background, the scheduler runs as an infinite loop that looks out for all the pods where the `nodeName` field is null. When it finds one, it starts the process of finding which node should be the best one. Though there are various parameters for it, just to name a few, they are `nodeName`, `nodeSelector`, taints, tolerations, `nodeAffinity`, and the rank of available nodes based on available resources in terms of CPU and memory, etc which decides the pod’s placement.

Author: ```
  - In the background, the scheduler runs an infinite loop looking for pods with null in the nodeName field.
  - When it finds one, it starts determining which node should be the best one for the pod.
  - There are various parameters for determining the pod's placement, but to name a few, they are `nodeName` , `nodeSelector` , `taints & tolerations` , `nodeAffinity` , rank of available nodes based on available resources in terms of CPU and memory, etc.
 ```
 Simplified the sentence.

- If in-case the scheduler is not able to find any node, the pod remains to be in pending state, which is not an ideal scenario.

The below diagram explains how the request flows in (starting from point number 1 to 4):

{{< figure src=kube-scheduler.svg alt="Whenever there is a new incoming pod, this is how kube-scheduler schedules a pod." title="Kube Scheduler">}}

## Typical Use Cases

> 

Reviewer 2: This section lacks a preface. You could start by saying that these are a few real-life scenarios where total control over the scheduling and eviction of pods is required. 

I'd recommend rewriting it entirely to incorporate a more generic overview since it delves way too much into specifics.

   - development [Priority 3]

   Unfortunately, the cluster goes through high resource consumption, and now there is a resource competition on the nodes with respect to the CPU and the memory. At least in the short term, you don’t have further nodes to expand or scale that Kubernetes cluster (cluster level autoscaling _may_ add more nodes, but this takes time). Perhaps some pods will remain in a Pending state, or perhaps service will degrade as Pods compete for resources. If the kubelet does evict a Pod from a node, that eviction happens at random because the kubelet doesn’t have any special information about which Pods to evict and which to keep.

3. Assume you have a microservices architecture that includes Kafka, Zookeeper, MongoDB, a couple of jobs, a couple of CronJobs, and around 40+ microservices. Assume that if we are caught up in this situation of resource crunch and somehow Kafka or MongoDB gets evicted, then all our microservices will be useless until and unless these pods are up and start serving traffic again. 

There can also be other scenarios where you want to control the order of scheduling or order of eviction of pods.

## PriorityClasses in Kubernetes

> PriorityClass is a cluster-wide API in Kubernetes, and is a part of the `scheduling.k8s.io/v1` API group. It contains a mapping of the PriorityClass name (defined in `.metadata.name`) and an integer value (defined in `.value`). This represents the value that the scheduler uses to determine Pods' relative priority.

Reviewer 2: ```suggestion
PriorityClass is a cluster-wide API object in Kubernetes and part of the `scheduling.k8s.io/v1` API group. It contains a mapping of the PriorityClass name (defined in `.metadata.name`) and an integer value (defined in `.value`). This represents the value that the scheduler uses to determine Pods' relative priority.
```

> Do you think Kubernetes itself uses PriorityClasses? The answer is yes. When you create a cluster using kubeadm, or even if you use a managed Kubernetes service (for example, Azure Kubernetes Service), PriorityClasses are widely used to safeguard the pods which are hosted on the control plane nodes. This aims to ensure that critical cluster components such as CoreDNS and kube-proxy are able to run even if resources are constrained.

Reviewer 2: ```suggestion
When you create a cluster using kubeadm, or even use a managed Kubernetes service (for example, Azure Kubernetes Service), Kubernetes uses PriorityClasses to safeguard the pods that are hosted on the control plane nodes. This ensures that critical cluster components such as CoreDNS and kube-proxy are able to run even if resources are constrained.
```

> Using a special PriorityClass helps to make sure that these pods are up and alive to serve the requests and make sure that the overall cluster is up and running.

Reviewer 2: ```suggestion
Using a special PriorityClass helps ensure that these pods can serve the requests and that the overall cluster is up and running.
```

```console
$ kubectl get priorityclass
NAME                      VALUE        GLOBAL-DEFAULT   AGE
system-cluster-critical   2000000000   false            82m
system-node-critical      2000001000   false            82m
```

> The below diagram shows exactly how it works and also contains the details of the example, which will come in the upcoming section (PriorityClass HandsOn Example).

Reviewer 2: ```suggestion
The diagram below shows exactly how it works and contains the example details, which will be detailed in the upcoming section.
```

{{< figure src=decision-tree.svg alt="Whenever there is a new incoming pod, this is how kube-scheduler decides their priority and puts them in a queue" title="Decision Tree">}}

### Pod priority and preemption

> Pod preemption is a Kubernetes feature that allows the cluster to [preempt](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption) lower-priority pods over high-priority pods in the event that there aren't enough resources to run all current Pods. To learn more about this concept, please refer to [Pod Priority and Preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/) in the Kubernetes documentation.

Reviewer 2: ```suggestion
[Pod preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption) is a Kubernetes feature that allows the cluster to preempt (evict) Pods on the basis of priority. Priority indicates the importance of a Pod relative to other Pods. The scheduler tries to evict lower priority Pods over high priority ones in the event that there aren't enough resources to run all current Pods. 
```

> Pod priority and preemption are Kubernetes features:

Reviewer 2: We're repeating ourselves here again without adding anything new for the first, second, and last bullet points. I suggest that we either retain the previous paragraph or expand this one to cover all the features of priority & premption.

Author: Sure, that make sense.

```
[Pod preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption) is a Kubernetes feature that allows the cluster to preempt (evict) pods on the basis of priority and [Pod priority](/docs/concepts/scheduling-eviction/pod-priority-preemption/#pod-priority) indicates the importance of a pod relative to other pods while scheduling. If there aren't enough resources to run all the current pods, the scheduler tries to evict lower-priority pods over high-priority ones.
```

Update the above paragraph and bullet points with above paragraph.

- Which help in deciding a priority while scheduling pods.
- Which check if the mission-critical pods are up and running and are not evicted.
- Decide how the cluster will behave in the event of a resource crunch.
- Decide which lower priority pods are preempted (evicted) from the nodes when there are resource crunch and high priority pods are available in the queue for scheduling.

### PriorityClass requirements

Before you set up PriorityClasses, there are a few things to consider.

> 1. Decide on which  PriorityClasses are needed.

Reviewer 2: I suggest we rephrase the sentence since these are user-specific considerations and not native to Kubernetes. From the examples, this isn't a "type" of PriorityClass but how a user could decide where it is used. Also, there is a spacing nit that I've addressed in the suggestion below.
```suggestion
1. Decide which PriorityClasses are needed.
```

Author: Sure.

   Example: Based on environments, or based on the kind of pods, or based on the kind of applications.
2. Which PriorityClass resource will be the default for the cluster (otherwise the pods without a `priorityClassName` will be treated as priority 0).
3. Respective PriorityClass resources in the cluster based on the decision made on #1.

> 4. How you'll name your priority classes.

Reviewer 2: Will this not be something that's decided as part of #1 when the user decides how the PriorityClass will be implemented within their setup?

Author: Reworded the bullet points:
```
1. Decide which PriorityClasses are needed.
For instance, based on environment, type of pods, type of applications, etc.
2. Which PriorityClass resource will be the default for the cluster (otherwise, the pods without a `priorityClassName` will be treated as priority 0).
3. Use a consistent naming convention for all PriorityClasses.
4. Make sure that the pods for your workloads are running with the right PriorityClass.
```

5. How you'll make sure that Pods for your workloads are running with the right PriorityClass.

## PriorityClass hands-on example

Let’s say there are 3 application pods: one for prod, one for preprod, and one for development. Below are three sample YAML manifest files for each of those.

```yaml
---
# development
apiVersion: v1
kind: Pod
metadata:
  name: dev-nginx
  labels:
    env: dev
spec:
  containers:
  - name: dev-nginx
    image: nginx
    resources:
      requests:
        memory: "256Mi"
         cpu: "0.2"
      limits:
        memory: ".5Gi"
        cpu: "0.5"
```

```yaml
---
# preproduction
apiVersion: v1
kind: Pod
metadata:
  name: preprod-nginx
  labels:
    env: preprod
spec:
  containers:
  - name: preprod-nginx
    image: nginx
    resources:
      requests:
        memory: "1.5Gi"
        cpu: "1.5"
      limits:
        memory: "2Gi"
        cpu: "2"
```

```yaml
---
# production
apiVersion: v1
kind: Pod
metadata:
  name: prod-nginx
  labels:
    env: prod
spec:
  containers:
  - name: prod-nginx
    image: nginx
    resources:
      requests:
        memory: "2Gi"
        cpu: "2"
      limits:
        memory: "2Gi"
        cpu: "2"
```

> Now, I'll create each of these pods by `kubectl create -f <FILE.yaml>` command, then check their status by using `kubectl get pods`. I can see if they are up and ready to serve traffic:

Reviewer 2: Kindly refrain from the usage of first-person pronouns.

Author: Sure, I will review this article again and will update it accordingly.

Author: Hi @divya-mohan0209 , 

I appreciate your time here, I have made changes based on all your recommendations.

Requesting you to please re-review the changes.

The preview version can be found [here.](https://deploy-preview-37713--kubernetes-io-main-staging.netlify.app/blog/2022/12/10/protect-mission-critical-pods-priorityclass/)

Please let me know if further changes are needed.

CC: @sftim 

Thank you,
Sunny
